### PR TITLE
feat: add CLI support

### PR DESCRIPTION
**Pull request by helio-frota**
* Adding CLI support using yargs
* Fix minor typos

### CODE REVIEW

**Comment by helio-frota** [comment]
Related to #12

**Review by wesleytodd — CHANGES_REQUESTED** [review]
Awesome work!  Thanks @helio-frota!

I think I am in favor of just taking positional args for this cli instead of a flag.  I would love to hear others thoughts on this before landing.

Just to illustrate the different approaches we might take:

```bash
$ nv lts
$ nv lts lts_latest
$ nv ~12.0.0

$ nv --lts
$ nv --lts --lts_latest
$ nv --semver="~12.0.0"


$ nv --show=lts
$ nv --show="lts || lts_latest"
$ nv --show="~12.0.0"
```

I am sure there are other ways we could consider, but with the above I think just taking positional arguments and passing them as an array to the `nv` call would be the most simple.

**Comment by wesleytodd** [comment]
cc @ljharb @dominykas @mhdawson @Eomm @dcodeIO (not sure who else to ping, we really need to sync the @nodejs/package-maintenance team to this org)

**Comment by ljharb** [comment]
I think it’s safer (keeps the design space open) to have positional args always be under a subcommand.

**Comment by helio-frota** [comment]
@wesleytodd thanks for the feedback. 
I'll wait the feedback of the others while I search how to have that working (positional) without a command.
The current published [doc](https://github.com/yargs/yargs/blob/master/docs/advanced.md) shows positional under a command like @ljharb wrote.

While I agree with @ljharb , we end up with `nv fooCommand lts` if I understood well. is that correct ?

**Comment by helio-frota** [comment]
interesting I just learned that `.usage` controls that positional behavior without commands

```console
14:27 $ node bin/nv.js a b c d e
nv.js [versions...]

Show Node.js versions

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
{ _: [ 'a', 'b', 'c', 'd', 'e' ], '$0': 'bin/nv.js' }
```

The problem is that this part is showing `nv.js [versions...]` which is not good looking  / seems that we need to type "versions" (or maybe it is only in my mind then)

@wesleytodd I like this approach: 

```
$ nv lts
$ nv lts lts_latest
$ nv ~12.0.0
```

But I have no idea yet how to get rid of `nv.js [versions...]`

This is the part of the code: 

```js
const args = require('yargs')
  .usage('$0 [versions...]', 'Show Node.js versions')
  .showHelp()
  .argv

if (args) {
  console.log(args)
...
```

**Comment by helio-frota** [comment]
Also this is working only when I add `.showHelp()` :man_shrugging:  

```
✔ ~/code/pkgjs/nv [add-cli-support L|✚ 1] 
14:36 $ node bin/nv.js ~12.0.0
[]
✔ ~/code/pkgjs/nv [add-cli-support L|✚ 1] 
14:36 $ node bin/nv.js ~12.0.0
nv.js [versions...]

Show Node.js versions

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
[
  {
    version: '12.0.0',
    major: 12,
    minor: 0,
    patch: 0,
    tag: '',
    codename: 'v12',
    versionName: 'v12',
    start: 2019-04-23T00:00:00.000Z,
    lts: 2019-10-21T00:00:00.000Z,
    maintenance: 2020-11-30T00:00:00.000Z,
    end: 2022-04-30T00:00:00.000Z
  }
]
```

**Comment by helio-frota** [comment]
able to put to work and tests are passing (locally)

```console
15:53 $ node bin/nv.js 8 9
[
  {
    version: '8.17.0',
    major: 8,
    minor: 17,
    patch: 0,
    tag: '',
    codename: 'carbon',
    versionName: 'v8',
    start: 2017-05-30T00:00:00.000Z,
    lts: 2017-10-31T00:00:00.000Z,
    maintenance: 2019-01-01T00:00:00.000Z,
    end: 2019-12-31T00:00:00.000Z
  },
  {
    version: '9.11.2',
    major: 9,
    minor: 11,
    patch: 2,
    tag: '',
    codename: 'v9',
    versionName: 'v9',
    start: 2017-10-01T00:00:00.000Z,
    lts: undefined,
    maintenance: 2018-04-01T00:00:00.000Z,
    end: 2018-06-30T00:00:00.000Z
  }
]
```

help is working with `-h`

```console
15:53 $ node bin/nv.js -h
nv.js [versions...]

Show Node.js versions

Options:
      --version  Show version number                                   [boolean]
  -h             Show help                                             [boolean]
```

But now when I try to run without positional arguments it shows node 12x : )

```console 
15:48 $ node bin/nv.js 
[
  {
    version: '12.19.0',
    major: 12,
    minor: 19,
    patch: 0,
    tag: '',
    codename: 'erbium',
    versionName: 'v12',
    start: 2019-04-23T00:00:00.000Z,
    lts: 2019-10-21T00:00:00.000Z,
    maintenance: 2020-11-30T00:00:00.000Z,
    end: 2022-04-30T00:00:00.000Z
  }
]
```

**Comment by wesleytodd** [comment]
`$ nv ls lts` maybe?

**Comment by helio-frota** [comment]
yeah `nv ls` seems cool. I think now it is a matter of implementation decision :+1:  , so please let me know then I update the PR.

**Comment by Eomm** [comment]
> While I agree with @ljharb , we end up with nv fooCommand lts if I understood well. is that correct ?

So would `ls` replace the `--show` option right?

And then the behaviour would be for example:

```sh
nv ls "10.0.0 || ~12.0.0"
```

Could we keep an alias to run `nv show "10.0.0 || ~12.0.0"` as well?

> Also this is working only when I add .showHelp()

Maybe [demandCommand](https://github.com/yargs/yargs/blob/master/docs/api.md#demandCommand) should help to show the help message when needed

**Comment by helio-frota** [comment]
> So would ls replace the --show option right?

yeah that's right

> Could we keep an alias to run nv show "10.0.0 || ~12.0.0" as well?

Maybe we can turn `show` from option to a command and not use `ls`  ?  : )

> Maybe demandCommand should help to show the help message when needed

Thanks, I think I tried that also but for some reason it was not worked : ( then I modified a bit:

`.usage('$0 [versions...]', 'Show Node.js versions', () => {})`  I added that extra parameter function `, () => {}` to get it working 

The code end up like this:

```js
const nv = require('../index')

const args = require('yargs')
  .usage('$0 [versions...]', 'Show Node.js versions', () => {})
  .help('h')
  .argv

nv(args.versions)
  .then(result => {
    console.log(result)
  }).catch(e => {
    console.error(e)
    process.exitCode = e.code || 1
  })
```

but again we still seeing this weird thing 

```
17:08 $ node bin/nv.js -h
nv.js [versions...]   <-------
```

because it is part of `.usage`

**Comment by helio-frota** [comment]
I was reading the documentation of [support](https://github.com/pkgjs/support/blob/master/doc/command-line-usage.md#package-consumers) and it has `show` command. Maybe is good to adopt the same `show` ?.

**Comment by dominykas** [comment]
> I think it’s safer (keeps the design space open) to have positional args always be under a subcommand.

In the spirit of small modules, do we need that design space for _this_ module? At the moment, it exports a single public function, so quite possibly just having the CLI take a single positional arg is good enough as well?

If we can agree we don't need subcommands, etc, I'd also question the need for yargs (`process.argv` works well enough for simple stuff?). 

Maybe we can keep this simple and then bump major if we have to? Or build a separate package for whatever the new needs might be.

**Comment by helio-frota** [comment]
I'm ok with process.argv and/or yargs
whatever is decided :+1:

**Comment by wesleytodd** [comment]
> In the spirit of small modules, do we need that design space for this module? At the moment, it exports a single public function, so quite possibly just having the CLI take a single positional arg (or essentially slurp all args) is good enough as well?

This was originally my thought as well, but I actually think `nv ls` is a pretty nice interface, so since it is good and simple it seems fine to leave ourselves open to the future changes.

> If we can agree we don't need subcommands, etc, I'd also question the need for yargs (process.argv works well enough for simple stuff?).

I disagree on this, usable help text is a requirement IMO.  Doing that with just `process.argv` is annoying at best. Additionally there is the ability to contribute when you use a framework like `yargs`, just learn it's well documented api.

> So would ls replace the --show option right?

That was my intent with the comment above.

> Maybe we can turn show from option to a command and not use ls ?

---

Once we agree on the interface we expose, I will also have comments on the usage of `yargs`, but I don't want to worry about those until we are sure of the api.

I am not as big of a fan of `show`, but am happy to have everyone else disagree and I will cave to the popular opinion.

**Comment by wesleytodd** [comment]
Hey, I am not sure we came to a consensus here.  My personal preference is that we go to either `nv show <versions>` or `nv ls <versions>`.  If noone opposes that after this comment for a day or so lets go with one of those.

I will go ahead and make detail comments on the `yargs` stuff soon (maybe layer today, I will leave this tab open).

**Comment by helio-frota** [comment]
PR updated, please let me know if that works :+1:

**Comment by ryami333** [comment]
Gidday folks, it's not obvious to me from reading the discussion or the PR diffs here whether this would support the use-case I suggested in this original issue, which is getting a _single_ result for a given release token:

```
$ npx @pkgjs/nv lts/erbium
12.19.0
```

Will this case be supported somehow? The application for this would be running this in a Github Workflow context, to set the node-version for `actions/setup-node`, like this:
```yml
echo '::set-output name=SELECTED_COLOR::green'

**Comment by ryami333** [comment]
Gidday folks, it's not obvious to me from reading the discussion or the PR diffs here whether this would support the use-case I suggested in this original issue, which is getting a _single_ result for a given release token:

```
$ npx @pkgjs/nv lts/erbium
12.19.0
```

Will this case be supported somehow? The application for this would be running this in a Github Workflow context, to set the node-version for `actions/setup-node`, like this:
```yml
    steps:
      - name: Checking out git
        uses: actions/checkout@master
      - name: Read .nvmrc
        run: echo ::set-output name=VERSION::$(npx @pkgjs/nv $(cat .nvmrc))
        id: nvm
      - name: Setup node
        uses: actions/setup-node@master
        with:
          node-version: "${{ steps.nvm.outputs.VERSION }}"

**Review by wesleytodd — CHANGES_REQUESTED** [review]
On a high-level, I prefer to expose my `yargs` based CLI's in the JS interface to allow for extensibility.  Typically this looks like:

```javascript
// cli.js
const yargs = require('yargs/yargs')
module.exports = (_yargs) => {
  const cli = _yargs || yargs()
  // ...
  return cli
}
```

```javascript
// bin/nv
require('../cli')().parse(process.argv.slice(2))
```


This makes it so folks who need to modify the behaviors or want to compose this cli into others can by simply passing their existing yargs instance.

One other thing I have found in the past is that on really wide terminals, the help output can look really weird, so I set a cap of 120 `cli.wrap(Math.min(cli.terminalWidth(), 120))`.  This is especially true for CLIs with very few options like this one.

File: bin/nv.js
Comment (helio-frota) [reply bin/nv.js]: partially resolving this. 
maybe it is good to have the now, cache and mirror options in another issue ?

File: test/cli.js
Comment (helio-frota) [reply test/cli.js]: same as previous comment to create an issue to add the extra options [?]

File: bin/nv.js
Comment (helio-frota) [reply bin/nv.js]: maybe it is good to create an issue for that also ?
maybe adding cli-table support or something ?

Comment (helio-frota) [reply bin/nv.js]: added line delimited output

File: bin/nv
Comment (ljharb) [inline bin/nv:14]:
```suggestion
            console.log(JSON.stringify(r, null, 2))
```

Comment (helio-frota) [reply bin/nv:14]: @ljharb yeah I remembered that I ran `dependents` other day and it outputs like this:
```
$ node bin/dependents.js --pkg roi
Getting first 10 dependents of roi sorted by downloads
{"tenorjs":{"downloads":169,"forks":4,"stars":4,"watchers":2,"url":"github.com/Jinzulen/TenorJS"}}
{"clojars-client":{"downloads":9,"forks":0,"stars":1,"watchers":0,"url":"github.com/panther-js/clojars-client"}}
...
```
and it has an option to output as json (which uses the null replacement and the space parameters).

So I was just guessing we should have the same behavior here but :man_shrugging: 

Other thing is I was suggesting @wesleytodd to move the remaining change-requests as issues so that we can discuss and maybe enable other people to contrib etc. 

Anyway. whatever you folks have a final decision I update the PR :+1:

**Comment by wesleytodd** [comment]
Closing in favor of #16

**Pull request closed without merging**
